tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+ * and represent RTT values for each server in a group. The RTT values
+ * are collected within a latest period of time (time window) and then
+ * used to predict the future RTT values that will be in action until
+ * the next run of this function. Server ratios are calculated on those
+ * predicted RTT values.
+ *
+ * A simple linear regression calculation on a sliding data window is
+ * used to predict future RTT values for each server. @y is an RTT value
+ * from APM, and @x is the current number of invocations of this timer
+ * function (every @intvl msecs). Essentially @x is a measure of time.
+ *
+ * The function runs periodically on timer and provides the data that
+ * is used by the ratio scheduler for outgoing requests.
+ *
+ * Return 0 if there are no new ratio values.
+ * Return a non-zero value if new ratio values were calculated.
+ */
+static int
+tfw_sched_ratio_calc_predict(TfwRatioPool *rpool, TfwRatio *ratio)
+{
+	TfwRatioHstData *hstdata = rpool->hstdata;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+	static const long MUL = 1000;
+	unsigned long x = hstdata->counter * MUL;
+	size_t si, sz, ni;
+
+	tfw_sched_ratio_fill_apmdata(rpool, ratio);
+
+	ni = hstdata->counter % hstdata->past_sz;
+
+	for (si = 0; si < rpool->srv_n; ++si) {
+		unsigned long y = srvdata[si].weight * MUL;
+		TfwRatioHstDesc *hd = &hstdata->past[si];
+
+		if (unlikely(hstdata->counter < hstdata->past_sz)) {
+			sz = ni + 1;
+			hd->x_avg = (hd->x_avg * ni + x) / sz;
+			hd->y_avg = (hd->y_avg * ni + y) / sz;
+			hd->xy_avg = (hd->xy_avg * ni + x * y) / sz;
+			hd->x_avg_y_avg = hd->x_avg * hd->y_avg;
+			hd->x_sq_avg = (hd->x_sq_avg * ni + x * x) / sz;
+			hd->x_avg_sq = hd->x_avg * hd->x_avg;
+		} else {
+			unsigned long h_x = hd->hist[ni].x;
+			unsigned long h_y = hd->hist[ni].y;
+			sz = hstdata->past_sz;
+			hd->x_avg = hd->x_avg - (h_x - x) / sz;
+			hd->y_avg = hd->y_avg - (h_y - y) / sz;
+			hd->xy_avg = hd->xy_avg - (h_x * h_y - x * y) / sz;
+			hd->x_avg_y_avg = hd->x_avg * hd->y_avg;
+			hd->x_sq_avg = hd->x_sq_avg - (h_x * h_x - x * x) / sz;
+			hd->x_avg_sq = hd->x_avg * hd->x_avg;
+		}
+
+		hd->hist[ni].x = x;
+		hd->hist[ni].y = y;
+
+		if (hd->x_sq_avg == hd->x_avg_sq) {
+			hd->a = 0;
+			hd->b = hd->x_avg ? hd->y_avg / hd->x_avg : 1;
+		} else {
+			hd->b = (hd->xy_avg - hd->x_avg_y_avg)
+				/ (hd->x_sq_avg - hd->x_avg_sq);
+			hd->a = (hd->y_avg - hd->b * hd->x_avg) / MUL;
+		}
+	}
+
+	x = hstdata->counter + hstdata->ahead;
+	for (si = 0; si < rpool->srv_n; ++si) {
+		TfwRatioHstDesc *hd = &hstdata->past[si];
+		long prediction = hd->a + hd->b * x;
+		srvdata[si].weight = prediction <= 0 ? 1 : prediction;
+	}
+	++hstdata->counter;
+
+	__tfw_sched_ratio_calc_dynamic(rpool, ratio);
+	return 1;
+}
+
+/**
+ * Get a free for use entry from the RCU pool.
+ */
+static TfwRatio *
+tfw_sched_ratio_rpool_get(TfwRatioPool *rpool)
+{
+	int si;
+	TfwRatio *ratio = rpool->rpool;
+
+	for (si = 0; si <= nr_cpu_ids; ++si, ++ratio) {
+		smp_mb();
```

Comment:
That's a good question. The initial idea was to pair atomic store/load operations on `ratio->free` so that the current state of the variable is visible on all CPUs.

Now that I look at it again it appears that all atomic operations on a specific `ratio->free` variable are always executed on the same CPU: The timer function calls `tfw_sched_ratio_rpool_get()` and `tfw_sched_ratio_rpool_put()`. It also calls `call_rcu()` which calls `tfw_sched_ratio_rpool_put()`. Then the callback is executed on the same CPU as the timer function is. That leads to the conclusion that memory barriers are not needed at all around `ratio-->free`. Is that a correct line of thought?

Also, perhaps this function should better be rewritten using cmpxchg? That would make it clearer and cleaner.